docs/proposals/scheduler/README.md

```diff
@@ -0,0 +1,135 @@
+* Server / Master: a server containing every administrative services (scheduler & controller)
+
+---
+
+## Architecture overview
+
+![Architecture overview](https://cdn.discordapp.com/attachments/828205813336244254/842476988290695187/a051f160-4d85-4c5c-a49b-0354ca9272d2.png)
+
+## Communication in the cluster 
+
+There are mainly two solutions. Either having direct connections between components (HTTP, gRPC...) of the cluster or having
+message queues where information is centralized.
+
+The second solution can be a great improvement for scalability and performance as everything would be completely asynchronous. 
+However, we need synchronous communication between internal services so make sure every information is properly received & understood.
+That's why the first solution direct communication between major components is a good choice. The drawback of this solution is we have 
+to write APIs for each component. 
+
+On top of that, we will be using [gRPC](https://grpc.io/) for communication between components. It will be handy to use 
+as API definitions are defined through [protoBuf](https://developers.google.com/protocol-buffers). 
+
+There is an interrogation around the controller & scheduler communication. They may be in the same 
+physical machine and not isolated one from another, so do we still need to use gRPC here ? Can't we 
+use any other solution of communication, as we are on the same physical machine ?
+
+The APIs exposed by our components must be defined through the need defined by the team 
+[controller](#controller) and the team [node](#node)
+
+---
+
+## Watcher 
+
+### Events 
+
+Watcher will work with an etcd to store usual and non highly dynamical data.
+The etcd will store:
+ - Number of workers
+ - workers alias (to improve/simplify communication)
+ - worker properties (such as cpu, RAM, memory, etc)
+
+Watcher is here to handle worker data, requesting it to the node manager/agent throught an API.
+Watcher will store in RAM the actual state of each nodes (idle, running, reloading, crashing, etc) and can give this metrics to scheduler when needed. If a node crash/restart, watcher will update datas in etcd.
+Watcher is the only point of communication with workers, it mean that he have to transfert scheduler instructions to the appropriate worker by resolving it using etcd alias name.
+
+### Recovery in case of crash
+
+Watcher self re-instanciate himself and get last data/metrics saved from etcd.
+etcd make sense because it provide a key/value storage that can be roles based ( leaders can read write update and other can readonly ). Provide safe cold storage on disk and great performance, can handle event on change (well to couple with watcher or controller maybe ). etcd is shared by all server components (read only), so we can get state metrics directly from controller or scheduler if it make sense.
+[source](https://www.ibm.com/cloud/learn/etcd)
+
```

Comment:
We discussed it, we decided to finally remove the name of `watcher`, and make it a part of the `scheduler` (so only the functions of the watcher stay). This is due because we will implement the push model you are talking about. Worker nodes will send metrics asynchronously. 

About the second point where the scheduler should be stateless, does it mean the whole cluster information should be saved in an external database (e.g. `etcd`). Or would the scheduler only have in memory the latest memory received from worker nodes ?

Reply:
> About the second point where the scheduler should be stateless, does it mean the whole cluster information should be saved in an external database (e.g. etcd). Or would the scheduler only have in memory the latest memory received from worker nodes ?

It means the scheduler should not need to keep any state permanently and also that it does not a given state to properly start. It can always start/restart from an empty list of workers and build an accurate description of all workers as they connect to the scheduler.